packages/alfa-table/src/table.ts

```diff
@@ -1,684 +1,854 @@
+     */
+    function ensure(x: number, y: number): void {
+      for (let i = table.length - 1; i < y; i++) {
+        table.push([]);
+      }
+
+      const row = table[y];
+
+      for (let i = row.length - 1; i < x; i++) {
+        row.push([]);
+      }
     }
 
-    public get colGroups(): Iterable<ColumnGroup> {
-      return this._table.columnGroups;
+    /**
+     * Get the cells at the given position.
+     */
+    function get(x: number, y: number): Array<number> {
+      ensure(x, y);
+      return table[y][x];
     }
 
-    public get rowGroups(): Iterable<RowGroup> {
-      return this._table.rowGroups;
+    /**
+     * Check if the table has cells at the given position.
+     */
+    function has(x: number, y: number): boolean {
+      return get(x, y).length > 0;
     }
 
-    public get table(): Table {
-      return Table.of(
-        this.element,
-        this.width,
-        this.height,
-        this._cells.map((cell) => cell.cell),
-        this.rowGroups,
-        this.colGroups
-      );
+    /**
+     * Add a cell at the given position.
+     */
+    function add(x: number, y: number, i: number): number {
+      if (cells[i].isData()) {
+        data.x.add(x);
+        data.y.add(y);
+      }
+
+      ensure(x, y);
+      return table[y][x].push(i);
     }
 
-    public slot(x: number, y: number): Option<Cell.Builder> {
-      return this._slots?.[x]?.[y] === undefined ? None : this._slots[x][y];
+    /**
+     * Get the current child element of the table.
+     */
+    function current(): Option<Element> {
+      return children.first();
     }
 
     /**
-     * Update by getting new values. Does not keep anything in sync, hence is highly unsafe. Use at your own risks.
+     * Advance to the next child element of the table.
      */
-    private _updateUnsafe({
-      element = this.element,
-      width = this.width,
-      height = this.height,
-      cells = this._cells,
-      downwardGrowingCells = this._downwardGrowingCells,
-      slots = this._slots,
-      rowGroups = this.rowGroups,
-      colGroups = this.colGroups,
-      rowGroupHeaders = this._rowGroupHeaders,
-      columnGroupHeaders = this._columnGroupHeaders,
-    }: {
-      element?: Element;
-      width?: number;
-      height?: number;
-      cells?: Iterable<Cell.Builder>;
-      downwardGrowingCells?: Iterable<Cell.Builder>;
-      slots?: Array<Array<Option<Cell.Builder>>>;
-      rowGroups?: Iterable<RowGroup>;
-      colGroups?: Iterable<ColumnGroup>;
-      rowGroupHeaders?: Iterable<Cell.Builder>;
-      columnGroupHeaders?: Iterable<Cell.Builder>;
-    }): Builder {
-      return Builder.of(
-        element,
-        width,
-        height,
-        cells,
-        downwardGrowingCells,
-        slots,
-        rowGroups,
-        colGroups,
-        rowGroupHeaders,
-        columnGroupHeaders
-      );
+    function advance(): Option<Element> {
+      children = children.rest();
```

Comment:
I'm not a huge fan of the global-ish `children` variable modified by side effect, this has potential to blow up in many ways (just like any global variable…)

Reply:
`fromTable()` defines a lot of mutable local variables, `children` included, as that aligns well with the way the algorithm is specified. It's all local to the function body though and is therefore by no means global. Trying to deviate from the specification to avoid mutable local variables seems like the greater evil to me, not the lesser.